bot/exts/utilities/bookmark.py

```diff
@@ -7,10 +7,77 @@
 
 from bot.bot import Bot
 from bot.constants import Colours, ERROR_REPLIES, Icons, Roles
+from bot.utils.converters import WrappedMessageConverter
 from bot.utils.decorators import whitelist_override
 
 log = logging.getLogger(__name__)
 
+MESSAGE_NOT_FOUND_ERROR = (
+    "You must either provide a reference to a valid message, or reply to one."
+    "\n\nThe lookup strategy for a message is as follows (in order):"
+    "\n1. Lookup by '{channel ID}-{message ID}' (retrieved by shift-clicking on 'Copy ID')"
+    "\n2. Lookup by message ID (the message **must** be in the current channel)"
+    "\n3. Lookup by message URL"
+)
+
+
+async def dm_bookmark(
+    target_user: discord.Member | discord.User,
+    target_message: discord.Message,
+    title: str,
+) -> None:
+    """
+    Sends the `target_message` as a bookmark to the `target_user` DMs, with `title` as the embed title..
+
+    Raises ``discord.Forbidden`` if the user's DMs are closed.
+    """
+    embed = Bookmark.build_bookmark_dm(target_message, title)
+    message_url_view = discord.ui.View().add_item(
+        discord.ui.Button(label="View Message", url=target_message.jump_url)
+    )
+    await target_user.send(embed=embed, view=message_url_view)
+    log.info(f"{target_user} bookmarked {target_message.jump_url} with title {title!r}")
+
+
+class SendBookmark(discord.ui.View):
+    """The button that sends the bookmark to other users."""
+
+    def __init__(
+        self,
+        author: discord.Member,
+        channel: discord.TextChannel,
+        target_message: discord.Message,
+        title: str,
+    ):
+        super().__init__()
+
+        self.clicked = [author.id]
+        self.channel = channel
+        self.target_message = target_message
+        self.title = title
+
+    @discord.ui.button(label="Receive Bookmark", style=discord.ButtonStyle.green)
+    async def button_callback(self, interaction: discord.Interaction, button: discord.ui.Button) -> None:
+        """The button callback."""
+        if interaction.user.id in self.clicked:
+            await interaction.response.send_message(
+                "You have already received a bookmark to that message.",
+                ephemeral=True,
+            )
+            return
+
+        try:
+            self.clicked.append(interaction.user.id)
```

Comment:
This shouldn't happen here. You should append only in the `else` block, once we've made sure that user has received the bookmark